From 2cf4eb878f07ae31682a20a9288c387f15420042 Mon Sep 17 00:00:00 2001
From: jiqichen <Jiqian.Chen@amd.com>
Date: Thu, 4 May 2023 10:56:09 +0800
Subject: [PATCH 2/2] seabios: perform pci device initialization also when
 running on xen

When running on XEN, the qemu platform initialization code does not call the
init functions of the enumerated pci devices.
This results in pci resume to not perform the necessary PIIX4 PM post-resume
setup, as it considers PIIX4 not present.

Signed-off-by: Xenia Ragiadakou <xenia.ragiadakou@amd.com>
---
 src/fw/paravirt.c  | 1 +
 src/fw/pciinit.c   | 2 +-
 src/hw/pcidevice.h | 1 +
 3 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/fw/paravirt.c b/src/fw/paravirt.c
index fba4e52d..0c035ebf 100644
--- a/src/fw/paravirt.c
+++ b/src/fw/paravirt.c
@@ -205,6 +205,7 @@ qemu_platform_setup(void)
 
     if (runningOnXen()) {
         pci_probe_devices();
+        pci_bios_init_devices();
         xen_hypercall_setup();
         xen_biostable_setup();
         return;
diff --git a/src/fw/pciinit.c b/src/fw/pciinit.c
index badf13d3..f9554811 100644
--- a/src/fw/pciinit.c
+++ b/src/fw/pciinit.c
@@ -417,7 +417,7 @@ static void pci_bios_init_device(struct pci_device *pci)
                          PCI_BRIDGE_CTL_SERR);
 }
 
-static void pci_bios_init_devices(void)
+void pci_bios_init_devices(void)
 {
     struct pci_device *pci;
     foreachpci(pci) {
diff --git a/src/hw/pcidevice.h b/src/hw/pcidevice.h
index 225d5451..fc807cba 100644
--- a/src/hw/pcidevice.h
+++ b/src/hw/pcidevice.h
@@ -63,6 +63,7 @@ struct pci_device_id {
     }
 
 void pci_probe_devices(void);
+void pci_bios_init_devices(void);
 struct pci_device *pci_find_device(u16 vendid, u16 devid);
 struct pci_device *pci_find_class(u16 classid);
 int pci_init_device(const struct pci_device_id *ids
-- 
2.17.1

